### ISSUE TITLE
Cannot resolve parent container IP address - with depends_on in config

### ISSUE THREAD

**Issue by lovmat** [docker/compose#2805]
I've an issue with the latest rc2 release and the depends_on feature. My config looks like this:
[docker-compose.yml](https://github.com/docker/compose/files/113802/docker-compose.yml.txt)

Getting error:
> front.example.org           | Message from VCC-compiler:
> front.example.org           | Backend host '"www.development.example.org"' could not be resolved to an IP address:
> front.example.org           | 	Name or service not known

So apparently the front.development.example.org container starts before it's able to resolve the host www.development.example.org to an IP address.
(I've attempted to manually sleep for two secs before starting front main application, and then it works fine.)

**Comment by aanand** [comment]
This sounds like a networking bug. Could you post the full output of `docker-compose --verbose up -d` so that we can confirm that the upstream container is created, started and connected to the network *before* the front container?

@madhu is it possible for a delay to occur between successfully connecting a container to a network and its name becoming resolvable?

**Comment by lovmat** [comment]
Sure:
[docker-compose-debug.txt](https://github.com/docker/compose/files/114032/docker-compose-debug.txt)

**Comment by aanand** [comment]
I see from those logs that the `www` service was already running (`compose.service.execute_convergence_plan: www.example.org is up-to-date`). Could you remove everything with `docker-compose down` first and then do a fresh run?

Also, perhaps it would be best to run without `-d` so we can see whether the DNS lookup succeeded. So just `docker-compose --verbose up`.

Finally, could you just paste the logs into your comment, surrounded by code fences (triple backticks, as [demonstrated here](https://guides.github.com/features/mastering-markdown/#examples) under "Code")? Thanks.

**Comment by lovmat** [comment]
Sorry, my bad. Here it goes:

```
$ docker-compose --verbose up
compose.config.config.find: Using configuration files: ./docker-compose.yml
docker.auth.auth.load_config: File doesn't exist
compose.cli.command.get_client: docker-compose version 1.6.0rc2, build 695c692
docker-py version: 1.7.0-rc3
CPython version: 2.7.9
OpenSSL version: OpenSSL 1.0.1e 11 Feb 2013
compose.cli.command.get_client: Docker base_url: http+docker://localunixsocket
compose.cli.command.get_client: Docker version: KernelVersion=4.2.0-25-generic, Os=linux, BuildTime=2016-01-27T22:35:52.038534916+00:00, ApiVersion=1.22, Version=1.10.0-rc2, GitCommit=c1cdc6e, Arch=amd64, GoVersion=go1.5.3
compose.cli.verbose_proxy.proxy_callable: docker containers <- (all=True, filters={u'label': [u'com.docker.compose.project=dockerexampleorg', u'com.docker.compose.service=www.development.example.org', u'com.docker.compose.oneoff=False']})
compose.cli.verbose_proxy.proxy_callable: docker containers -> (list with 1 items)
compose.cli.verbose_proxy.proxy_callable: docker inspect_container <- (u'462a74dbd1227f712e8dd53be5be8436c16616a290da63535f6a77f6cec8b314')
compose.cli.verbose_proxy.proxy_callable: docker inspect_container -> {u'AppArmorProfile': u'',
 u'Args': [],
 u'Config': {u'AttachStderr': False,
             u'AttachStdin': False,
             u'AttachStdout': False,
             u'Cmd': [u'/sbin/my_init'],
             u'Domainname': u'',
             u'Entrypoint': None,
             u'Env': [u'PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin',
                      u'DEBIAN_FRONTEND=noninteractive'],
...
compose.cli.verbose_proxy.proxy_callable: docker containers <- (all=True, filters={u'label': [u'com.docker.compose.project=dockerexampleorg', u'com.docker.compose.service=front.development.example.org', u'com.docker.compose.oneoff=False']})
compose.cli.verbose_proxy.proxy_callable: docker containers -> (list with 1 items)
compose.cli.verbose_proxy.proxy_callable: docker inspect_container <- (u'281d0ba61edb8320d37e4723c959af2787785cdf8d05d56995d2ba4ca6fb7656')
compose.cli.verbose_proxy.proxy_callable: docker inspect_container -> {u'AppArmorProfile': u'',
 u'Args': [],
 u'Config': {u'AttachStderr': False,
             u'AttachStdin': False,
             u'AttachStdout': False,
             u'Cmd': None,
             u'Domainname': u'',
             u'Entrypoint': [u'/entrypoint.sh'],
             u'Env': [u'affinity:container==a32050bb4739bd03b65a265b18c7ef3b63a4225748c9e50330e7c93ac375cc44',
                      u'PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin',
...
compose.cli.verbose_proxy.proxy_callable: docker containers <- (all=True, filters={u'label': [u'com.docker.compose.project=dockerexampleorg', u'com.docker.compose.service=www.development.example.org', u'com.docker.compose.oneoff=False']})
compose.cli.verbose_proxy.proxy_callable: docker containers -> (list with 1 items)
compose.cli.verbose_proxy.proxy_callable: docker inspect_image <- (u'dockerexampleorg_www.development.example.org')
compose.cli.verbose_proxy.proxy_callable: docker inspect_image -> {u'Architecture': u'amd64',
 u'Author': u'',
 u'Comment': u'',
 u'Config': {u'ArgsEscaped': True,
             u'AttachStderr': False,
             u'AttachStdin': False,
             u'AttachStdout': False,
             u'Cmd': [u'/sbin/my_init'],
             u'Domainname': u'',
             u'Entrypoint': None,
...
compose.cli.verbose_proxy.proxy_callable: docker inspect_container <- (u'462a74dbd1227f712e8dd53be5be8436c16616a290da63535f6a77f6cec8b314')
compose.cli.verbose_proxy.proxy_callable: docker inspect_container -> {u'AppArmorProfile': u'',
 u'Args': [],
 u'Config': {u'AttachStderr': False,
             u'AttachStdin': False,
             u'AttachStdout': False,
             u'Cmd': [u'/sbin/my_init'],
             u'Domainname': u'',
             u'Entrypoint': None,
             u'Env': [u'PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin',
                      u'DEBIAN_FRONTEND=noninteractive'],
...
compose.cli.verbose_proxy.proxy_callable: docker containers <- (all=True, filters={u'label': [u'com.docker.compose.project=dockerexampleorg', u'com.docker.compose.service=front.development.example.org', u'com.docker.compose.oneoff=False']})
compose.cli.verbose_proxy.proxy_callable: docker containers -> (list with 1 items)
compose.cli.verbose_proxy.proxy_callable: docker inspect_image <- (u'dockerexampleorg_front.development.example.org')
compose.cli.verbose_proxy.proxy_callable: docker inspect_image -> {u'Architecture': u'amd64',
 u'Author': u'',
 u'Comment': u'',
 u'Config': {u'ArgsEscaped': True,
             u'AttachStderr': False,
             u'AttachStdin': False,
             u'AttachStdout': False,
             u'Cmd': None,
             u'Domainname': u'',
             u'Entrypoint': [u'/entrypoint.sh'],
...
compose.cli.verbose_proxy.proxy_callable: docker inspect_container <- (u'281d0ba61edb8320d37e4723c959af2787785cdf8d05d56995d2ba4ca6fb7656')
compose.cli.verbose_proxy.proxy_callable: docker inspect_container -> {u'AppArmorProfile': u'',
 u'Args': [],
 u'Config': {u'AttachStderr': False,
             u'AttachStdin': False,
             u'AttachStdout': False,
             u'Cmd': None,
             u'Domainname': u'',
             u'Entrypoint': [u'/entrypoint.sh'],
             u'Env': [u'affinity:container==a32050bb4739bd03b65a265b18c7ef3b63a4225748c9e50330e7c93ac375cc44',
                      u'PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin',
...
compose.cli.verbose_proxy.proxy_callable: docker inspect_network <- (u'dockerexampleorg_default')
compose.cli.verbose_proxy.proxy_callable: docker inspect_network -> {u'Containers': {},
 u'Driver': u'bridge',
 u'IPAM': {u'Config': [{u'Gateway': u'172.18.0.1/16',
                        u'Subnet': u'172.18.0.0/16'}],
           u'Driver': u'default',
           u'Options': None},
 u'Id': u'b7c011b7364a9ffdeda4b2373fa24d82e8433ec8126a65b3d499b89f1330b23b',
 u'Name': u'dockerexampleorg_default',
 u'Options': {},
 u'Scope': u'local'}
compose.service.start_container_if_stopped: Starting www.example.org
compose.cli.verbose_proxy.proxy_callable: docker attach <- (u'462a74dbd1227f712e8dd53be5be8436c16616a290da63535f6a77f6cec8b314', stderr=True, stream=True, stdout=True)
compose.cli.verbose_proxy.proxy_callable: docker attach -> <generator object _multiplexed_response_stream_helper at 0x7faa41d49d70>
compose.cli.verbose_proxy.proxy_callable: docker start <- (u'462a74dbd1227f712e8dd53be5be8436c16616a290da63535f6a77f6cec8b314')
compose.cli.verbose_proxy.proxy_callable: docker start -> None
compose.cli.verbose_proxy.proxy_callable: docker disconnect_container_from_network <- (u'462a74dbd1227f712e8dd53be5be8436c16616a290da63535f6a77f6cec8b314', u'dockerexampleorg_default')
compose.cli.verbose_proxy.proxy_callable: docker disconnect_container_from_network -> None
compose.cli.verbose_proxy.proxy_callable: docker connect_container_to_network <- (u'462a74dbd1227f712e8dd53be5be8436c16616a290da63535f6a77f6cec8b314', u'dockerexampleorg_default', links=[], aliases=['www.development.example.org', u'462a74dbd1'])
compose.cli.verbose_proxy.proxy_callable: docker connect_container_to_network -> None
compose.service.start_container_if_stopped: Starting front.example.org
compose.cli.verbose_proxy.proxy_callable: docker attach <- (u'281d0ba61edb8320d37e4723c959af2787785cdf8d05d56995d2ba4ca6fb7656', stderr=True, stream=True, stdout=True)
compose.cli.verbose_proxy.proxy_callable: docker attach -> <generator object _multiplexed_response_stream_helper at 0x7faa41a681e0>
compose.cli.verbose_proxy.proxy_callable: docker start <- (u'281d0ba61edb8320d37e4723c959af2787785cdf8d05d56995d2ba4ca6fb7656')
compose.cli.verbose_proxy.proxy_callable: docker start -> None
compose.cli.verbose_proxy.proxy_callable: docker disconnect_container_from_network <- (u'281d0ba61edb8320d37e4723c959af2787785cdf8d05d56995d2ba4ca6fb7656', u'dockerexampleorg_default')
compose.cli.verbose_proxy.proxy_callable: docker disconnect_container_from_network -> None
compose.cli.verbose_proxy.proxy_callable: docker connect_container_to_network <- (u'281d0ba61edb8320d37e4723c959af2787785cdf8d05d56995d2ba4ca6fb7656', u'dockerexampleorg_default', links=[], aliases=['front.development.example.org', u'281d0ba61e'])
compose.cli.verbose_proxy.proxy_callable: docker connect_container_to_network -> None
Attaching to www.example.org, front.example.org
www.example.org   | *** Running /etc/my_init.d/00_apache2...
front.example.org | Message from VCC-compiler:
front.example.org | Backend host '"www.development.example.org"' could not be resolved to an IP address:
front.example.org | 	Name or service not known
```

**Comment by aanand** [comment]
OK, it definitely looks like everything's happening in the right order, so I'd guess this is a bug with the DNS resolver set up by libnetwork.

**Comment by aanand** [comment]
Just realised I meant to ping @mavenugo above, not madhu.

**Comment by mavenugo** [comment]
@aanand thanks. there is nothing obvious in libnetwork that could cause this delay.
The embedded DNS should be ready to go when the container starts.

The logs indicate some activity regarding `disconnect_container_from_network` for an ID and `None` followed by connect. Why is that done ? It seems to be done after the container has started. Why is that ?

**Comment by aanand** [comment]
One of the aliases we want to set for the container is a short id, which we don't know until the container has been created, so we can't set that alias at creation time by putting it in `NetworkSettings`. Instead, we disconnect it from the network and reconnect it with the alias included.

**Comment by mavenugo** [comment]
@aanand sure. but can you do the `disconnect` and `connect` before the start ?
That will guarantee that the network plumbing is done (including DNS) before the application starts.

**Comment by aanand** [comment]
Yep, looks like the race is in Compose. Fix in #2808.

**Comment by aficiomaquinas** [comment]
I'm having the same problem on Docker Cloud. Does Docker Cloud has a different mechanism to do this? When I ran it locally it worked perfectly, but deploying it on cloud using a stackfile does not seem to work.
